### ISSUE TITLE
NGL WebGL protein viewer support

### ISSUE THREAD

**Issue by Jhsmit** [holoviz/panel#1307]
#### Is your feature request related to a problem? Please describe.
I'd like to incorporate a protein viewer in my panel application. To do so I'm looking at [ngl](https://github.com/arose/ngl) which has both an ipywidget for notebooks and JS/HTML available for embedding.

#### Describe the solution you'd like
I'd like to include the protein viewer as a panel in my existing app, and control properties of this viewer from python and/or other widgets. 

#### Describe alternatives you've considered

I can serve the following HTML using python's ``http.server``:

```
<!DOCTYPE html>
<html lang="en">
<head>
  <title>NGL - embedded</title>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, user-scalable=no, minimum-scale=1.0, maximum-scale=1.0">
</head>
<body>
  <script src="https://cdn.jsdelivr.net/gh/arose/ngl@v2.0.0-dev.33/dist/ngl.js"></script>
  <script>
    var stage;
    document.addEventListener("DOMContentLoaded", function () {
      stage = new NGL.Stage("viewport");
      stage.loadFile("rcsb://1NKT.mmtf", {defaultRepresentation: true});
    });
  </script>
  <div id="viewport" style="width:800px; height:800px;"></div>
</body>
</html>
```

I've tried to incorporate this HTML by using ``pn.pane.HTML``, but this gives me the errors ``Exceeded 16 live WebGL contexts for this principal, losing the least recently used one.``
``WebGL context was lost.```

So I've adapted this to:
```
<!DOCTYPE html>
<html lang="en">
<head>
  <title>NGL - embedded</title>
  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, user-scalable=no, minimum-scale=1.0, maximum-scale=1.0">
</head>
<body>
  <script>
    var stage;
    document.addEventListener("DOMContentLoaded", function () {
      stage = new NGL.Stage("viewport");
      stage.loadFile("rcsb://1NKT.mmtf", {defaultRepresentation: true});
    });
  </script>
  <div id="viewport" style="width:800px; height:800px;"></div>
</body>
</html>
```
and loaded the JS through pn.extension. This doesnt give any errors but doenst render anything either.

Using the ipywidget [nglview](https://github.com/arose/nglview) I can use the viewer in jupyter notebooks:

```python
import nglview
view = nglview.show_pdbid("3pqr")  # load "3pqr" from RCSB PDB and display viewer widget
view
```

I've tried to put this in a panel:

```python
import panel as pn
import nglview

js_files = {'ngl': 'https://cdn.jsdelivr.net/gh/arose/ngl@v2.0.0-dev.33/dist/ngl.js'}
pn.extension(comms='ipywidgets', js_files=js_files)

view = nglview.show_pdbid("3pqr")  
pane = pn.Pane(view)

pane
```
In a jupyter notebook this gives:

```
Error: Script error for "@bokeh/jupyter_bokeh"
http://requirejs.org/docs/errors.html#scripterror
```

In jupyter lab I get as output:

```NGLWidget(n_components=1)``` 

I do have the most recent version of ``jupyter_bokeh`` installed.

In summary, I'm wondering how to do this and if its already possible with the current version or if some adaptation is needed.

@MarcSkovMadsen Is it possible to do this with the ``WebComponent``? Or can I somehow get the nglview ipywidget working?

**Comment by MarcSkovMadsen** [comment]
Hi @Jhsmit 

This is a truly amazing library. I think you just need to change `NGL` to `ngl`. At least this works

```python
import panel as pn
pn.config.js_files["ngl"]="https://cdn.jsdelivr.net/gh/arose/ngl@v2.0.0-dev.33/dist/ngl.js"
pn.extension()

html = """<div id="viewport" style="width:100%; height:100%;"></div>
<script>
stage = new ngl.Stage("viewport");
stage.loadFile("rcsb://1NKT.mmtf", {defaultRepresentation: true});
</script>"""


ngl_pane = pn.pane.HTML(html, height=500, width=500)
ngl_pane
```

![image](https://user-images.githubusercontent.com/42288570/80495191-e2ec5f80-8967-11ea-85c6-b56191d4a408.png)

**Comment by MarcSkovMadsen** [comment]
And yes. When the WebComponent gets deployed you should be able to setup integrations to selected attributes, properties and events pretty quickly.

**Comment by philippjfr** [comment]
We're still working on the last edge cases with ipywidgets support, that said `jupyter_bokeh` is not the library you want anyway. You want `ipywidgets_bokeh`.

**Comment by MarcSkovMadsen** [comment]
I posted the code in Discourse as well because others might search there.

https://discourse.holoviz.org/t/how-to-use-ngl-webgl-protein-viewer-in-panel/583

**Comment by Jhsmit** [comment]
@philippjfr 

When I use ``ipywidgets_bokeh``

```python
import panel as pn
from ipywidgets_bokeh import IPyWidget
import nglview


view = nglview.show_pdbid("3pqr")
wrapper = IPyWidget(widget=view, width=800, height=800)

pane = pn.panel(wrapper)

pane.servable()
```
It doesnt quite work.
I have some missing fonts (I think) 

``` 
2020-04-28 18:47:31,522 404 GET /static/extensions/ipywidgets_bokeh/af7ae505a9eed503f8b8e6982036873e.woff2 (::1) 0.00ms
2020-04-28 18:47:31,554 404 GET /static/extensions/ipywidgets_bokeh/fee66e712a8a08eef5805a46892932ad.woff (::1) 0.00ms
```
And:

```
TypeError: "buffers is undefined"
    widgets 2.7.5/dist/index.js:116602
    _handleCommMsg http://localhost:5006/static/extensions/ipywidgets_bokeh/ipywidgets_bokeh.js:69
    _handleMessage http://localhost:5006/static/extensions/ipywidgets_bokeh/ipywidgets_bokeh.js:69
ipywidgets_bokeh.js:69:43739
    _msgChain http://localhost:5006/static/extensions/ipywidgets_bokeh/ipywidgets_bokeh.js:69
```

However, it does do something because I can see the fullscreen and the resize buttons (and they work).
Should I make an issue for this over at ``ipywidgets_bokeh``?

**Comment by philippjfr** [comment]
>Should I make an issue for this over at ipywidgets_bokeh?

I think so yes.

**Comment by Jhsmit** [comment]
@MarcSkovMadsen 
Thanks, but for some reason this seems to only works in jupyter notebook.

If I run this example with ``.servable()`` and ``panel serve`` I do need ngl to be with capital letters (``NGL.Stage("viewport"))

This library is indeed very nice and I'm looking forward to having WebComponents and interacting with it! 

Maybe I could make an example for this library at some point

**Comment by philippjfr** [comment]
This is now covered by https://github.com/MarcSkovMadsen/panel-chemistry
